### PR TITLE
Update gophish complete

### LINKED ISSUE #2: Clean the structure of `gophish-complete`
# 🚀 Feature Proposal

Improve the clarity of the `gophish-complete` tool by modifying the options `--auto` and `--complete` into a single option with an optional `CAMPAIGN_ID`. The difference between these flags  is minimal and can be combined.

## Motivation

This will make it easier to understand the function of the commands and simplify the code. 

## Example

Get rid of the `--auto` option
Support an optional `CAMPAIGN_ID` for both `--complete` and `--summary`
So the docopt line would look something like:

`gophish-complete (--complete [CAMPAIGN_ID] | --summary [CAMPAIGN_ID]) [--log-level=LEVEL] SERVER API_KEY`

## Pitch

This is based on a suggestion/feedback from @dav3r

### PR DESCRIPTION
**Pull request by bjb28**
# 

Modified `gophish-complete` to have improved command line interaction and use the campaign name as an argument. 

## 🗣 Description

To make the command line interaction easier this PR removes the use of the flag `auto` for passing a campaign identifier to the script. The new method takes the command name as an optional argument which triggers the script to complete the campaign with that name. 

Additionally, this PR changes from using the campaign id to campaign name for automatic actions. 

## 💭 Motivation and Context

This change cleans up the command line interface making it easier to understand. It also alters to the use of campaign name which will be a known element at any point in the assessment. If the campaign id was still used, the process requires knowing an identifier that is automatically set by GoPhish. 

This PR would resolve #2.

## 🧪 Testing

A Pytest was created for the `get_campaign_id` that tests a successful and unsuccessful lookup. 

Conducted the following manual tests using `gophish-complete` and a GoPhish server:
- used both `--summary` and `--complete` to ensure the user was prompted for an assessment id followed by the user selecting a campaign from a list.
- used `CAMPAIN_NAME` with `--summary` and `--complete` to ensure the desired campaign was completed or summarized.
- Both tests above were conducted with non-existent assessments and campaigns to ensure the expected errors were produced. 

## 🚥 Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (causes existing functionality to change)

## ✅ Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/tools/gophish_complete.py [hunk 1]
```diff
@@ -3,14 +3,15 @@
 """Complete a campaign in GoPhish and/or output a GoPhish campaign summary.
 
 Usage:
-  gophish-complete (--auto=CAMPAIGN_ID | --complete | --summary ) [--log-level=LEVEL] SERVER API_KEY
+  gophish-complete (--complete | --summary) [--log-level=LEVEL] SERVER API_KEY
+  gophish-complete CAMPAIGN_NAME (--complete | --summary) [--log-level=LEVEL] SERVER API_KEY
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: Can't we consolidate these two docopt lines into a single line like this?
```
gophish-complete (--complete | --summary) [--campaign=NAME] [--log-level=LEVEL] SERVER API_KEY
```

File: src/tools/gophish_complete.py [hunk 2]
```diff
@@ -3,14 +3,15 @@
...
   gophish-complete (-h | --help)
   gophish-complete --version
 
 Options:
   API_KEY                   GoPhish API key.
+  CAMPAIGN_NAME             GoPhish campaign name.
   SERVER                    Full URL to GoPhish server.
-  -a --auto=CAMPAIGN_ID     Complete a specific campaign identified by CAMPAIGN_ID.
   -c --complete             Display a list of campaigns to choose which to complete.
   -s --summary              Choose a campaign from a list to output a summary.
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: The option descriptions for `--complete` and `--summary` should be updated to only mention their function- there should not be any mention about displaying a list of campaigns or choosing a campaign, since that is now going to depend on whether or not the optional campaign name has been provided.

File: src/tools/gophish_complete.py [hunk 3]
```diff
@@ -46,46 +48,66 @@
...
-    """Return a dictionary containing all campaigns."""
+def get_campaign_id(campaign_name, campaigns):
+    """Get campaign id from campaign name.
+
+    Args:
+        campaign_name (string): Full campaign name.
+        campaigns (dict): Campaign id as key, campaign name as value.
+
+    Raises:
+        LookupError: When the campaign name is not found, raise exception.
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: I'd simplify this wording slightly to: `Campaign name is not found in campaigns dictionary.`

File: src/tools/gophish_complete.py [hunk 4]
```diff
@@ -46,46 +48,66 @@
...
+
+    Return:
+        Return campign id that corresponds to campaign name provided.
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: Match verb tense with rest of docstring and eliminate redundant words:
```
Returns:
    Campaign id corresponding to campaign name provided.
```

File: src/tools/gophish_complete.py [hunk 5]
```diff
@@ -46,46 +48,66 @@
...
+    """
+    for campaign_id, name_value in campaigns.items():
+        if name_value == campaign_name:
+            return campaign_id
+
+    raise LookupError(f'Campaign Name "{campaign_name}" not found.')
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: `Campaign Name` -> `Campaign name`

File: src/tools/gophish_complete.py [hunk 6]
```diff
@@ -46,46 +48,66 @@
...
+
+
+def get_campaigns(api, assessment_id):
+    """Return a dictionary containing all campaigns.
+
+    Args:
+        api (GoPhish API): Connection to GoPhish server via the API
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: Missing a period at the end of this sentence.

File: src/tools/gophish_complete.py [hunk 7]
```diff
@@ -46,46 +48,66 @@
...
+        assessment_id (string): Assessment identifier to get campaigns from.
+
+    Raises:
+        LookupError: If no campaigns are found for the assessment identifier, raise exception.
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: `LookupError: No campaigns found for the provided assessment id.`

File: src/tools/gophish_complete.py [hunk 8]
```diff
@@ -46,46 +48,66 @@
...
 
-        if len(assessmentCampaigns):
-            break
-        else:
-            logging.warning(
-                "No Campaigns found for Assessment {}".format(assessment_id)
-            )
-            print("Please try again...")
+    if len(assessmentCampaigns) == 0:
+        raise LookupError(f"No Campaigns found for Assessment {assessment_id}")
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: `No Campaigns found for Assessment` -> `No campaigns found for assessment`

File: src/tools/gophish_complete.py [hunk 9]
```diff
@@ -137,35 +159,60 @@ def main():
...
-        workingID = select_campaigns(campaigns)
-        success = print_summary(api, workingID)
-
-    elif args["--auto"]:
-        success = complete_campaign(
-            api, args["API_KEY"], args["SERVER"], args["--auto"]
-        )
+        # Sets assessment id from first section of campaign name. If the
+        # assessment wizard is used to build the assessment the campaign
+        # name will allways start with the assessment identifier.
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: Typo: `allways` -> `always`

File: src/tools/gophish_complete.py (hunk 2 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Fixed in commit 1ecf246.

File: src/tools/gophish_complete.py [hunk 10]
```diff
@@ -137,35 +159,60 @@ def main():
...
+        assessment_id = args["CAMPAIGN_NAME"].split("-")[0]
+
+    # Gather all campaigns associated with assessment identifier.
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: If the campaign name is provided, why do we need to derive the assessment_id and fetch all campaigns from that assessment?  Can't we just find the requested campaign directly from a dict of all campaigns?

File: src/tools/gophish_complete.py [hunk 11]
```diff
@@ -3,24 +3,24 @@
...
+  -s --summary              Output a summary of a GoPhish campaign.
   -h --help                 Show this screen.
   --version                 Show version.
   -l --log-level=LEVEL      If specified, then the log level will be set to
                             the specified value.  Valid values are "debug", "info",
                             "warning", "error", and "critical". [default: info]
 
 NOTE:
-  * If the auto flag is not provided, all assessment campaigns will be listed to select from.
+  * If the campaign option with a campaign name is not provided, all assessment campaigns will be listed to select from.
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: I think you can simply say `If a campaign name is not provided, all...`

File: src/tools/gophish_complete.py (hunk 1 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Great idea! Implemented this change in commit fa471ad. 

Successfully tested with both `-C NAME` and `--Campaign=NAME` to ensure the desired campaign was used.  Also verified running without the `--Campaign` option did not impact the manual process.

File: src/tools/gophish_complete.py [hunk 12]
```diff
@@ -3,24 +3,24 @@
...
   gophish-complete (-h | --help)
   gophish-complete --version
 
 Options:
   API_KEY                   GoPhish API key.
   SERVER                    Full URL to GoPhish server.
-  -a --auto=CAMPAIGN_ID     Complete a specific campaign identified by CAMPAIGN_ID.
-  -c --complete             Display a list of campaigns to choose which to complete.
-  -s --summary              Choose a campaign from a list to output a summary.
+  -C --Campaign=NAME        GoPhish campaign name.
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: We don't usually do capital letter arguments.  As I think about this more, our args can be further simplified:

```
gophish-complete [--campaign=NAME] [--summary-only] [--log-level=LEVEL] SERVER API_KEY
```
The default behavior will be to complete a campaign, so no need to have the `--complete` arg at all.  The optional behavior will be triggered via `--summary-only`.  What do you think?

File: src/tools/gophish_complete.py [hunk 13]
```diff
@@ -46,46 +47,66 @@
...
 
-def select_campaigns(campaigns):
+def select_campaign(campaigns):
     """Return the ID of a selected campaign."""
     print("Please select a Campaign ID:")
     print("\tID: Name")
 
     for key, name in campaigns.items():
-        print("\t" + str(key) + ": " + name)
+        print(f"\t {str(key)}: {name}")
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: I don't think you still need the `str()` here now that this is an f-string, but please confirm that.

Also, I'd prefer the item variables here to be `id, name` to match the language used elsewhere in this code.

File: tests/test_tools.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    )
+    def test_get_campaign_id_found(self, campaigns):
+        """Test finding campaign id by campaign name when successful."""
+        assert get_campaign_id("RV0000-C2", campaigns) == "2"
+
+    @pytest.mark.parametrize(
+        "campaigns", [{"1": "RV0000-C1", "2": "RV0000-C2", "3": "RV0000-C3"}]
+    )
+    def test_get_campaign_id_not_found(self, campaigns):
+        """Test getting campaign id bu campaign name when unsuccessful."""
```
Comment (dav3r) [inline tests/test_tools.py]: I'd reword this a bit:  `"""Verify LookupError is raised when searching for unknown campaign id."""`

File: tests/test_tools.py [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+
+
+class TestComplete:
+    """Test gophis-complete script."""
+
+    @pytest.mark.parametrize(
+        "campaigns", [{"1": "RV0000-C1", "2": "RV0000-C2", "3": "RV0000-C3"}]
+    )
+    def test_get_campaign_id_found(self, campaigns):
+        """Test finding campaign id by campaign name when successful."""
```
Comment (dav3r) [inline tests/test_tools.py]: I'd reword this a bit:  `"""Verify correct campaign id is returned when a valid campaign name is provided."""`

**Review by dav3r — CHANGES_REQUESTED** [review]
I've got a bunch of comments in here- some minor, some that will require a little more work.  I didn't review the rest of the changes towards the bottom of `main()` because I suspect they will change now- I will review them on the next pass.

File: src/tools/gophish_complete.py (hunk 10 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Good point, and I think this would be achievable.

File: src/tools/gophish_complete.py (hunk 13 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Fixed in commit 1ab0c81.

File: src/tools/gophish_complete.py (hunk 11 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Fixed in commit 1ab0c81.

File: src/tools/gophish_complete.py (hunk 3 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Fixed in commit 1ab0c81.

File: src/tools/gophish_complete.py (hunk 4 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Fixed in commit 1ab0c81.

File: src/tools/gophish_complete.py (hunk 5 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Fixed in commit 1ab0c81.

File: src/tools/gophish_complete.py (hunk 6 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Fixed in commit 1ab0c81.

File: src/tools/gophish_complete.py (hunk 7 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Fixed in commit 1ab0c81.

File: src/tools/gophish_complete.py (hunk 8 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Fixed in commit 1ab0c81.

File: src/tools/gophish_complete.py (hunk 9 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Fixed in commit 1ab0c81.

File: src/tools/gophish_complete.py (hunk 12 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: As discussed, I think this is a good idea. I am working on making the needed changes.

File: src/tools/gophish_complete.py (hunk 10 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Made this change in commit 59ca3fb. The script sets the `assessment_id` to a blank string so all campaigns will be returned and search for the campaign name.

File: src/tools/gophish_complete.py (hunk 12 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Changes made to adjust the functionality in commit 9658acd. 

I also adjusted `complete_campaign` so it will raise a warning if GoPhish returns the completion as unsuccessful.  This change allowed me to remove many of the conditional statements and repetitive setting of `success`.

File: tests/test_tools.py (hunk 14 above)
Comment (bjb28) [reply tests/test_tools.py]: Changed wording in commit 517e478.

File: tests/test_tools.py (hunk 15 above)
Comment (bjb28) [reply tests/test_tools.py]: Changed wording in commit 517e478.

File: src/tools/gophish_complete.py [hunk 16]
```diff
@@ -46,46 +46,66 @@
...
+    """Get campaign id from campaign name.
+
+    Args:
+        campaign_name (string): Full campaign name.
+        campaigns (dict): Campaign id as key, campaign name as value.
+
+    Raises:
+        LookupError: Campaign name is not found in campaigns dictionary.
+
+    Return:
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: `Return:` --> `Returns:`  😃

File: src/tools/gophish_complete.py [hunk 17]
```diff
@@ -94,33 +114,41 @@ def select_campaigns(campaigns):
     return inputId
 
 
-def complete_campaign(api, api_key, server, workingID):
-    """Complete a campaign."""
+def complete_campaign(api_key, server, workingID):
```
Comment (dav3r) [inline src/tools/gophish_complete.py]: As long as we're cleaning things up in here, can you please change `workingID` to `campaign_id`?

File: tests/test_tools.py [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+
+# Third-Party Libraries
+import pytest
+
+# cisagov Libraries
+from tools.gophish_complete import get_campaign_id
+
+
+class TestComplete:
+    """Test gophis-complete script."""
```
Comment (dav3r) [inline tests/test_tools.py]: Typo: `gophis-complete` -> `gophish-complete`

Comment (bjb28) [reply tests/test_tools.py]: Correction made in commit b9dbc4f.

File: src/tools/gophish_complete.py (hunk 16 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Corrected in commit b9dbc4f.

File: src/tools/gophish_complete.py (hunk 17 above)
Comment (bjb28) [reply src/tools/gophish_complete.py]: Changed variable in commit 6e2412c.

Comment (dav3r) [reply src/tools/gophish_complete.py]: I still see `workingID` variable in a few places:  https://github.com/bjb28/gophish-tools/blob/update-gophish-complete/src/tools/gophish_complete.py#L121-L146

Also, missing a period at the end of this line: https://github.com/bjb28/gophish-tools/blob/update-gophish-complete/src/tools/gophish_complete.py#L127

Comment (bjb28) [reply src/tools/gophish_complete.py]: Corrected these items in cbc0c60 and used find to double-check this time. :mag:

Comment (dav3r) [reply src/tools/gophish_complete.py]: Looks good now!

**Review by dav3r — APPROVED** [review]
Sweet- everything is looking ship-shape now!  Thanks a ton for making all of these updates!  🚀
